src/algebra/group/hom.lean

```diff
@@ -184,15 +243,199 @@ attribute [instance] inv.is_group_hom neg.is_add_group_hom
+attribute [to_additive add_monoid_hom.mk] monoid_hom.mk
+attribute [to_additive add_monoid_hom.mk.inj] monoid_hom.mk.inj
+attribute [to_additive add_monoid_hom.mk.inj_arrow] monoid_hom.mk.inj_arrow
+attribute [to_additive add_monoid_hom.mk.inj_eq] monoid_hom.mk.inj_eq
+attribute [to_additive add_monoid_hom.mk.sizeof_spec] monoid_hom.mk.sizeof_spec
+attribute [to_additive add_monoid_hom.no_confusion] monoid_hom.no_confusion
+attribute [to_additive add_monoid_hom.no_confusion_type] monoid_hom.no_confusion_type
+attribute [to_additive add_monoid_hom.rec] monoid_hom.rec
+attribute [to_additive add_monoid_hom.rec_on] monoid_hom.rec_on
+attribute [to_additive add_monoid_hom.sizeof] monoid_hom.sizeof
+attribute [to_additive add_monoid_hom.to_fun] monoid_hom.to_fun
+
+namespace monoid_hom
+variables {M : Type*} {N : Type*} {P : Type*} [monoid M] [monoid N] [monoid P]
+variables {G : Type*} {H : Type*} [group G] [comm_group H]
+
+@[extensionality] def ext (f g : M →* N) (h : (f : M → N) = g) : f = g :=
+by cases f; cases g; cases h; refl
+
+/-- If f is a monoid homomorphism then f 1 = 1. -/
+@[simp] lemma map_one (f : M →* N) : f 1 = 1 := f.map_one'
+
+/-- If f is a monoid homomorphism then f (a * b) = f a * f b. -/
+@[simp] lemma map_mul (f : M →* N) (a b : M) : f (a * b) = f a * f b := f.map_mul' a b
+
+/-- The identity map from a monoid to itself. -/
+def id (M : Type*) [monoid M] : M →* M :=
+{ to_fun := id,
+  map_one' := rfl,
+  map_mul' := λ _ _, rfl }
+
+/-- Composition of monoid morphisms is a monoid morphism. -/
+def comp (hnp : N →* P) (hmn : M →* N) : M →* P :=
+{ to_fun := hnp ∘ hmn,
+  map_one' := by simp,
+  map_mul' := by simp,
+}
+
+/-- The product of two monoids morphisms is a monoid morphism if the target is commutative. -/
+def mul {M N} [monoid M] [comm_monoid N] (f g : M →* N) : M →* N :=
+{ to_fun := λ m, f m * g m,
+  map_one' := show f 1 * g 1 = 1, by simp,
+  map_mul' := begin intros, show f (x * y) * g (x * y) = f x * g x * (f y * g y),
+    rw [f.map_mul, g.map_mul, ←mul_assoc, ←mul_assoc, mul_right_comm (f x)], end
+}
+
+/-- Group homomorphisms preserve inverse. -/
+theorem map_inv {G H} [group G] [group H] (f : G →* H) (g : G) : f g⁻¹ = (f g)⁻¹ :=
+eq_inv_of_mul_eq_one $ by rw [←f.map_mul, inv_mul_self, f.map_one]
+
+/-- Makes a group homomomorphism from a proof that the map preserves multiplication. -/
+def mk' {f : M → G} (map_mul : ∀ a b : M, f (a * b) = f a * f b) : M →* G :=
+{ to_fun := f,
+  map_mul' := map_mul,
+  map_one' := mul_self_iff_eq_one.1 $ by rw [←map_mul, mul_one]
+}
+
+/-- Makes a group homomorphism from a map and a proof that it preserves multiplication. -/
+def mk'' (f : M → G) (map_mul : ∀ a b : M, f (a * b) = f a * f b) : M →* G := mk' map_mul
+
+/-- The inverse of a monoid homomorphism is a monoid homomorphism if the target is
+    a commutative group.-/
+def inv {M H} [monoid M] [comm_group H] (f : M →* H) : M →* H :=
+mk'' (λ g, (f g)⁻¹) $ λ a b, by rw [←mul_inv, f.map_mul]
+
```

Comment:
Do you want to put a `monoid` instance on `G →* H`? So that we can just write `f * g`? And extend this to `group` when `H` is a commutative group?
Also, at the moment people can't use any of the libraries that we have, because they all assume `is_monoid_hom` or `is_group_hom`. To make the transition possible, I think we should have an `is_monoid_hom` instance for the the coerced `f : G → H` if `f : G →* H`. (And an `is_group_hom` if `G` and `H` are groups.)

Reply:
Monoid instance: I've put comm_monoid instance on M -> N when the target is a comm_monoid and a comm_group instance when the target is a comm_group (and all the additive versions). I've made the coercion functions instances as well.